### PR TITLE
add Hetero Factorization Machine

### PR DESCRIPTION
**Pull request by peiyong86**
Fixes ISSUE https://github.com/FederatedAI/FATE/issues/958

Changes:

1. Add federatedRec module.

2. Add Hetero FM to federatedRec.

### CODE REVIEW

File: federatedrec/factorization_machine/hetero_factorization_machine/hetero_fm_arbiter.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+            if self.model_param.early_stop == 'weight_diff':
+                weight_diff = fate_operator.norm(total_gradient)
+                LOGGER.info("iter: {}, weight_diff:{}, is_converged: {}".format(self.n_iter_,
+                                                                                weight_diff, self.is_converged))
+                if weight_diff < self.model_param.tol:
+                    self.is_converged = True
+            else:
+                if iter_loss is None:
+                    raise ValueError("Multiple host situation, loss early stop function is not available."
```
Comment (mgqa34) [inline federatedrec/factorization_machine/hetero_factorization_machine/hetero_fm_arbiter.py]: A little confusion about this line, HeteroFM does not support multi-host, but  but changes "early stop" to weight diff will support multi-host?

File: federatedrec/factorization_machine/README.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
+# Federated Factorization Machine
+
+Factorization Machine(FM) is a  supervised learning approach incorporates second-order feature interaction.
+Federated factorization machine computes these cross-party cross-features and their gradients under encryption. 
+
+Here we simplify participants of the federation process into three parties. Party A represents Guest, party B represents Host. Party C, which is also known as “Arbiter,” is a third party that works as coordinator. Party C is responsible for generating private and public keys.
+
+## Heterogeneous FM
```
Comment (mgqa34) [inline federatedrec/factorization_machine/README.md]: The readme here can't show the federated processing, can some references be attached to here ot the image can show more ? Like encryption and gradient exchanges

File: federatedrec/factorization_machine/hetero_factorization_machine/hetero_fm_base.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        super()._init_model(params)
+        self.encrypted_mode_calculator_param = params.encrypted_mode_calculator_param
+        self.cipher_operator = PaillierEncrypt()
+        self.transfer_variable = HeteroFMTransferVariable()
+        self.cipher.register_paillier_cipher(self.transfer_variable)
+        self.converge_procedure.register_convergence(self.transfer_variable)
+        self.batch_generator.register_batch_generator(self.transfer_variable)
+        self.gradient_loss_operator.register_gradient_procedure(self.transfer_variable)
+
+    def transform(self, data_inst):
```
Comment (mgqa34) [inline federatedrec/factorization_machine/hetero_factorization_machine/hetero_fm_base.py]: Does this function really need here ? In HeteroLR, this function is called by dnn_lr, but it seems like that federated recommendation will not be invoked that?

File: federatedrec/param/factorization_machine_param.py [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+
+
+class FactorizationParam(BaseParam):
+    """
+    Parameters used for Factorization Machine.
+
+    Parameters
+    ----------
+    penalty : str, 'L1' or 'L2'. default: 'L2'
+        Penalty method used in LR. Please note that, when using encrypted version in HomoLR,
```
Comment (mgqa34) [inline federatedrec/param/factorization_machine_param.py]: Here "LR" shoulb modify to "FM"

File: federatedrec/factorization_machine/hetero_factorization_machine/hetero_fm_arbiter.py (hunk 1 above)
Comment (peiyong86) [reply federatedrec/factorization_machine/hetero_factorization_machine/hetero_fm_arbiter.py]: Removed this check. Thanks.

File: federatedrec/factorization_machine/hetero_factorization_machine/hetero_fm_base.py (hunk 3 above)
Comment (peiyong86) [reply federatedrec/factorization_machine/hetero_factorization_machine/hetero_fm_base.py]: Removed transform. Thanks.

File: federatedrec/param/factorization_machine_param.py (hunk 4 above)
Comment (peiyong86) [reply federatedrec/param/factorization_machine_param.py]: Fixed. Thanks.

File: federatedrec/factorization_machine/README.md (hunk 2 above)
Comment (peiyong86) [reply federatedrec/factorization_machine/README.md]: Thanks for remind. Will check with the author, and update here as soon as possible.

**Pull request merged**